File: app/src/main/java/org/connectbot/util/GnomeTerminalColorSchemeParser.java
/*
 * ConnectBot: simple, powerful, open-source SSH client for Android
 * Copyright 2026
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.util;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Parses common GNOME Terminal color scheme exports into a 16-color ANSI palette.
 *
 * <p>Supported inputs (best-effort):</p>
 * <ul>
 *   <li>{@code dconf dump} / {@code gsettings get} style entries:
 *     <ul>
 *       <li>{@code palette=['rgb(...)', ...]}</li>
 *       <li>{@code foreground-color='rgb(...)'}</li>
 *       <li>{@code background-color='rgb(...)'}</li>
 *     </ul>
 *   </li>
 *   <li>Common "Gogh" theme scripts:
 *     <ul>
 *       <li>{@code PALETTE="#RRGGBB:#RRGGBB:... (16 entries)"}</li>
 *       <li>{@code FOREGROUND_COLOR="#RRGGBB"}, {@code BACKGROUND_COLOR="#RRGGBB"}</li>
 *     </ul>
 *   </li>
 * </ul>
 */
public final class GnomeTerminalColorSchemeParser {
	private GnomeTerminalColorSchemeParser() {
	}

	public static final class ParsedScheme {
		public final int[] ansiColors;
		public final Integer foregroundColor;
		public final Integer backgroundColor;

		private ParsedScheme(int[] ansiColors, Integer foregroundColor, Integer backgroundColor) {
			this.ansiColors = ansiColors;
			this.foregroundColor = foregroundColor;
			this.backgroundColor = backgroundColor;
		}
	}

	private static final Pattern ASSIGNMENT =
			Pattern.compile("^\\s*(?:export\\s+)?([A-Za-z0-9_-]+)\\s*=\\s*(.*?)\\s*$");

	private static final Pattern QUOTED_TOKEN = Pattern.compile("['\"]([^'\"]+)['\"]");

	private static final Pattern RGB =
			Pattern.compile("(?i)rgba?\\(\\s*([0-9.]+)\\s*,\\s*([0-9.]+)\\s*,\\s*([0-9.]+)\\s*(?:,\\s*([0-9.]+)\\s*)?\\)");

	private static final Pattern COLON_HEX_PALETTE =
			Pattern.compile("(?i)#[0-9a-f]{6}(?::#[0-9a-f]{6}){15}");

	public static boolean looksLikeGnomeTerminalScheme(String text) {
		if (text == null) return false;
		return text.toLowerCase(Locale.US).contains("palette") || COLON_HEX_PALETTE.matcher(text).find();
	}

	public static ParsedScheme parse(String text) {
		if (text == null) throw new IllegalArgumentException("text is null");

		Map<String, String> assignments = parseAssignments(text);

		String paletteValue = assignments.get("palette");
		int[] palette =
				paletteValue != null ? parsePaletteValue(paletteValue) : parsePaletteFallback(text);
		if (palette == null) {
			throw new IllegalArgumentException("No GNOME Terminal palette found");
		}

		Integer fg = null;
		Integer bg = null;
		String fgRaw = firstNonNull(assignments.get("foreground_color"), assignments.get("fg_color"));
		String bgRaw = firstNonNull(assignments.get("background_color"), assignments.get("bg_color"));
		if (fgRaw != null) fg = parseColorString(fgRaw);
		if (bgRaw != null) bg = parseColorString(bgRaw);

		return new ParsedScheme(palette, fg, bg);
	}

	private static Map<String, String> parseAssignments(String text) {
		Map<String, String> result = new LinkedHashMap<>();
		String[] lines = text.split("\n");
		for (String rawLine : lines) {
			if (rawLine == null) continue;
			String line = rawLine.trim();
			if (line.isEmpty()) continue;
			if (line.startsWith("#") || line.startsWith(";")) continue;

			Matcher m = ASSIGNMENT.matcher(rawLine);
			if (!m.matches()) continue;
			String key = normalizeKey(m.group(1));
			String value = m.group(2) != null ? m.group(2).trim() : "";
			result.put(key, value);
		}
		return result;
	}

	private static String normalizeKey(String key) {
		if (key == null) return "";
		return key.trim().toLowerCase(Locale.US).replace('-', '_');
	}

	private static String firstNonNull(String a, String b) {
		return a != null ? a : b;
	}

	private static int[] parsePaletteFallback(String text) {
		Matcher m = COLON_HEX_PALETTE.matcher(text);
		if (m.find()) {
			return parseColonSeparatedPalette(m.group());
		}
		return null;
	}

	private static int[] parsePaletteValue(String value) {
		String unwrapped = stripOuterQuotes(value.trim()).trim();
		if (unwrapped.endsWith(";")) unwrapped = unwrapped.substring(0, unwrapped.length() - 1).trim();

		if (unwrapped.startsWith("[")) {
			return parseBracketListPalette(unwrapped);
		}
		if (unwrapped.contains(":")) {
			return parseColonSeparatedPalette(unwrapped);
		}
		throw new IllegalArgumentException("Unsupported palette format");
	}

	private static int[] parseBracketListPalette(String listText) {
		String inner = listText.trim();
		if (inner.startsWith("[")) inner = inner.substring(1);
		if (inner.endsWith("]")) inner = inner.substring(0, inner.length() - 1);
		inner = inner.trim();
		if (inner.isEmpty()) throw new IllegalArgumentException("Empty palette list");

		List<String> tokens = new ArrayList<>();
		Matcher quoted = QUOTED_TOKEN.matcher(inner);
		while (quoted.find()) {
			tokens.add(quoted.group(1));
		}
		if (tokens.isEmpty()) {
			String[] parts = inner.split(",");
			for (String part : parts) {
				String t = part.trim();
				if (!t.isEmpty()) tokens.add(t);
			}
		}

		if (tokens.size() != 16) {
			throw new IllegalArgumentException("Expected 16 palette colors, got " + tokens.size());
		}

		int[] palette = new int[16];
		for (int i = 0; i < 16; i++) {
			palette[i] = parseColorString(tokens.get(i));
		}
		return palette;
	}

	private static int[] parseColonSeparatedPalette(String value) {
		String unwrapped = stripOuterQuotes(value.trim()).trim();
		String[] parts = unwrapped.split(":");
		List<String> tokens = new ArrayList<>();
		for (String p : parts) {
			String t = p.trim();
			if (!t.isEmpty()) tokens.add(t);
		}
		if (tokens.size() != 16) {
			throw new IllegalArgumentException("Expected 16 palette colors, got " + tokens.size());
		}
		int[] palette = new int[16];
		for (int i = 0; i < 16; i++) {
			palette[i] = parseColorString(tokens.get(i));
		}
		return palette;
	}

	private static String stripOuterQuotes(String value) {
		if (value == null) return "";
		String trimmed = value.trim();
		if (trimmed.length() < 2) return trimmed;
		char first = trimmed.charAt(0);
		char last = trimmed.charAt(trimmed.length() - 1);
		if ((first == '\'' && last == '\'') || (first == '"' && last == '"')) {
			return trimmed.substring(1, trimmed.length() - 1);
		}
		return trimmed;
	}

	private static int parseColorString(String value) {
		String cleaned = stripOuterQuotes(value == null ? "" : value.trim()).trim();
		if (cleaned.isEmpty()) throw new IllegalArgumentException("Empty color");

		Matcher rgb = RGB.matcher(cleaned);
		if (rgb.matches()) {
			double r = Double.parseDouble(rgb.group(1));
			double g = Double.parseDouble(rgb.group(2));
			double b = Double.parseDouble(rgb.group(3));
			double a = rgb.group(4) != null ? Double.parseDouble(rgb.group(4)) : 1.0;

			int rr = toByte(normalizeComponent(r));
			int gg = toByte(normalizeComponent(g));
			int bb = toByte(normalizeComponent(b));
			int aa = toByte(normalizeAlpha(a));
			return (aa << 24) | (rr << 16) | (gg << 8) | bb;
		}

		return parseHexColor(cleaned);
	}

	private static double normalizeComponent(double value) {
		if (Double.isNaN(value)) return 0.0;
		if (value <= 0.0) return 0.0;
		if (value <= 1.0) return value;
		if (value <= 255.0) return value / 255.0;
		return 1.0;
	}

	private static double normalizeAlpha(double value) {
		if (Double.isNaN(value)) return 1.0;
		if (value <= 0.0) return 0.0;
		if (value <= 1.0) return value;
		if (value <= 255.0) return value / 255.0;
		return 1.0;
	}

	private static int toByte(double v) {
		int b = (int) Math.round(v * 255.0);
		if (b < 0) return 0;
		if (b > 255) return 255;
		return b;
	}

	private static int parseHexColor(String value) {
		String trimmed = value.trim();
		String hex = trimmed;
		if (hex.startsWith("#")) hex = hex.substring(1);
		if (hex.startsWith("0x") || hex.startsWith("0X")) hex = hex.substring(2);
		hex = hex.toLowerCase(Locale.US);

		if (hex.length() == 3) {
			hex = "" + hex.charAt(0) + hex.charAt(0) + hex.charAt(1) + hex.charAt(1) + hex.charAt(2) + hex.charAt(2);
		} else if (hex.length() == 4) {
			hex = "" + hex.charAt(0) + hex.charAt(0) + hex.charAt(1) + hex.charAt(1) + hex.charAt(2) + hex.charAt(2) + hex.charAt(3) + hex.charAt(3);
		}

		if (hex.length() == 6) {
			int r = Integer.parseInt(hex.substring(0, 2), 16);
			int g = Integer.parseInt(hex.substring(2, 4), 16);
			int b = Integer.parseInt(hex.substring(4, 6), 16);
			return (0xFF << 24) | (r << 16) | (g << 8) | b;
		}
		if (hex.length() == 8) {
			boolean looksLikeAarrggbb = hex.startsWith("ff") || hex.startsWith("00");
			if (looksLikeAarrggbb) {
				int a = Integer.parseInt(hex.substring(0, 2), 16);
				int r = Integer.parseInt(hex.substring(2, 4), 16);
				int g = Integer.parseInt(hex.substring(4, 6), 16);
				int b = Integer.parseInt(hex.substring(6, 8), 16);
				return (a << 24) | (r << 16) | (g << 8) | b;
			} else {
				int r = Integer.parseInt(hex.substring(0, 2), 16);
				int g = Integer.parseInt(hex.substring(2, 4), 16);
				int b = Integer.parseInt(hex.substring(4, 6), 16);
				int a = Integer.parseInt(hex.substring(6, 8), 16);
				return (a << 24) | (r << 16) | (g << 8) | b;
			}
		}
		throw new IllegalArgumentException("Unsupported color format: " + value);
	}
}

File: app/src/main/java/org/connectbot/data/GnomeTerminalColorSchemeParser.kt
/*
 * ConnectBot: simple, powerful, open-source SSH client for Android
 * Copyright 2026
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.data

import kotlin.math.roundToInt

/**
 * Parses common GNOME Terminal color-scheme exports into a terminal palette.
 *
 * Supported inputs (best-effort):
 * - `dconf dump` / `gsettings get` style entries:
 *   - `palette=['rgb(...)', ...]`
 *   - `foreground-color='rgb(...)'`
 *   - `background-color='rgb(...)'`
 * - Common "Gogh" theme scripts:
 *   - `PALETTE="#RRGGBB:#RRGGBB:... (16 entries)"`
 *   - `FOREGROUND_COLOR="#RRGGBB"`, `BACKGROUND_COLOR="#RRGGBB"`
 *
 * Note: GNOME Terminal stores foreground/background as arbitrary colors, but ConnectBot stores defaults
 * as indices into the 16-color palette; callers may map FG/BG to closest ANSI entries.
 */
object GnomeTerminalColorSchemeParser {

    data class ParsedScheme(
        val ansiColors: IntArray,
        val foregroundColor: Int?,
        val backgroundColor: Int?,
    )

    fun looksLikeGnomeTerminalScheme(text: String): Boolean =
        text.contains("palette", ignoreCase = true) || COLON_HEX_PALETTE_REGEX.containsMatchIn(text)

    fun parse(text: String): ParsedScheme {
        val assignments = parseAssignments(text)

        val paletteValue = assignments[KEY_PALETTE]
        val palette =
            when {
                paletteValue != null -> parsePaletteValue(paletteValue)
                else -> parsePaletteFallback(text)
            } ?: throw IllegalArgumentException("No GNOME Terminal palette found")

        val foreground =
            assignments[KEY_FOREGROUND_COLOR]?.let { parseColorString(it) }
                ?: assignments[KEY_FG_COLOR]?.let { parseColorString(it) }
        val background =
            assignments[KEY_BACKGROUND_COLOR]?.let { parseColorString(it) }
                ?: assignments[KEY_BG_COLOR]?.let { parseColorString(it) }

        return ParsedScheme(
            ansiColors = palette,
            foregroundColor = foreground,
            backgroundColor = background,
        )
    }

    private const val KEY_PALETTE = "palette"
    private const val KEY_FOREGROUND_COLOR = "foreground_color"
    private const val KEY_BACKGROUND_COLOR = "background_color"
    private const val KEY_FG_COLOR = "fg_color"
    private const val KEY_BG_COLOR = "bg_color"

    private val ASSIGNMENT_REGEX =
        Regex("""^\s*(?:export\s+)?([A-Za-z0-9_-]+)\s*=\s*(.*?)\s*$""")

    private fun normalizeKey(key: String): String =
        key.trim().lowercase().replace('-', '_')

    private fun parseAssignments(text: String): Map<String, String> {
        val result = LinkedHashMap<String, String>()
        for (rawLine in text.lineSequence()) {
            val line = rawLine.trim()
            if (line.isBlank()) continue
            if (line.startsWith("#") || line.startsWith(";")) continue

            val match = ASSIGNMENT_REGEX.matchEntire(rawLine) ?: continue
            val key = normalizeKey(match.groupValues[1])
            val value = match.groupValues[2].trim()
            result[key] = value
        }
        return result
    }

    private fun parsePaletteFallback(text: String): IntArray? {
        BRACKET_LIST_REGEX.findAll(text).forEach { match ->
            val candidate = match.value
            try {
                return parseBracketListPalette(candidate)
            } catch (_: Exception) {
                // Ignore and keep searching
            }
        }

        val colonPalette = COLON_HEX_PALETTE_REGEX.find(text)?.value
        if (colonPalette != null) {
            try {
                return parseColonSeparatedPalette(colonPalette)
            } catch (_: Exception) {
                // Ignore
            }
        }

        return null
    }

    private val BRACKET_LIST_REGEX = Regex("""\[[^\]]+\]""", RegexOption.DOT_MATCHES_ALL)
    private val COLON_HEX_PALETTE_REGEX = Regex(
        """(?i)#[0-9a-f]{6}(?::#[0-9a-f]{6}){15}"""
    )

    private fun parsePaletteValue(value: String): IntArray {
        val unwrapped = stripOuterQuotes(value.trim()).trim().removeSuffix(";").trim()
        return when {
            unwrapped.startsWith("[") -> parseBracketListPalette(unwrapped)
            unwrapped.startsWith("(") -> parseParenListPalette(unwrapped)
            unwrapped.contains(":") -> parseColonSeparatedPalette(unwrapped)
            else -> throw IllegalArgumentException("Unsupported palette format")
        }
    }

    private fun parseBracketListPalette(listText: String): IntArray {
        val inner = listText.trim().removeSurrounding("[", "]").trim()
        if (inner.isBlank()) throw IllegalArgumentException("Empty palette list")

        val quotedTokens = QUOTED_TOKEN_REGEX.findAll(inner).map { it.groupValues[1] }.toList()
        val rawTokens =
            if (quotedTokens.isNotEmpty()) {
                quotedTokens
            } else {
                inner.split(',').map { it.trim() }.filter { it.isNotBlank() }
            }

        val colors = rawTokens.map { parseColorString(it) }
        if (colors.size != 16) throw IllegalArgumentException("Expected 16 palette colors, got ${colors.size}")
        return colors.toIntArray()
    }

    private fun parseParenListPalette(listText: String): IntArray {
        val inner = listText.trim().removeSurrounding("(", ")").trim()
        if (inner.isBlank()) throw IllegalArgumentException("Empty palette list")
        val tokens = QUOTED_TOKEN_REGEX.findAll(inner).map { it.groupValues[1] }.toList()
        val rawTokens =
            if (tokens.isNotEmpty()) tokens else inner.split(Regex("""\s+""")).map { it.trim() }.filter { it.isNotBlank() }
        val colors = rawTokens.map { parseColorString(it) }
        if (colors.size != 16) throw IllegalArgumentException("Expected 16 palette colors, got ${colors.size}")
        return colors.toIntArray()
    }

    private fun parseColonSeparatedPalette(value: String): IntArray {
        val unwrapped = stripOuterQuotes(value.trim()).trim()
        val parts = unwrapped.split(':').map { it.trim() }.filter { it.isNotBlank() }
        if (parts.size != 16) throw IllegalArgumentException("Expected 16 palette colors, got ${parts.size}")
        return parts.map { parseColorString(it) }.toIntArray()
    }

    private val QUOTED_TOKEN_REGEX = Regex("""['"]([^'"]+)['"]""")

    private fun stripOuterQuotes(value: String): String {
        val trimmed = value.trim()
        if (trimmed.length < 2) return trimmed
        val first = trimmed.first()
        val last = trimmed.last()
        return if ((first == '\'' && last == '\'') || (first == '"' && last == '"')) {
            trimmed.substring(1, trimmed.length - 1)
        } else {
            trimmed
        }
    }

    private fun parseColorString(value: String): Int {
        val cleaned = stripOuterQuotes(value.trim()).trim()
        if (cleaned.isBlank()) throw IllegalArgumentException("Empty color")

        val rgbMatch = RGB_REGEX.matchEntire(cleaned)
        if (rgbMatch != null) {
            val r = rgbMatch.groupValues[1].toDouble()
            val g = rgbMatch.groupValues[2].toDouble()
            val b = rgbMatch.groupValues[3].toDouble()
            val a = rgbMatch.groupValues.getOrNull(4)?.takeIf { it.isNotBlank() }?.toDouble() ?: 1.0

            return argbInt(
                r = normalizeComponent(r),
                g = normalizeComponent(g),
                b = normalizeComponent(b),
                a = normalizeAlpha(a),
            )
        }

        return parseHexColor(cleaned)
    }

    private val RGB_REGEX =
        Regex("""(?i)rgba?\(\s*([0-9.]+)\s*,\s*([0-9.]+)\s*,\s*([0-9.]+)\s*(?:,\s*([0-9.]+)\s*)?\)""")

    private fun normalizeComponent(value: Double): Double =
        when {
            value.isNaN() -> 0.0
            value <= 0.0 -> 0.0
            value <= 1.0 -> value
            value <= 255.0 -> value / 255.0
            else -> 1.0
        }

    private fun normalizeAlpha(value: Double): Double =
        when {
            value.isNaN() -> 1.0
            value <= 0.0 -> 0.0
            value <= 1.0 -> value
            value <= 255.0 -> value / 255.0
            else -> 1.0
        }

    private fun parseHexColor(value: String): Int {
        val trimmed = value.trim()
        val hex =
            trimmed.removePrefix("#")
                .removePrefix("0x")
                .removePrefix("0X")
                .lowercase()

        val expanded =
            when (hex.length) {
                3 -> "${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}"
                4 -> "${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}${hex[3]}${hex[3]}"
                else -> hex
            }

        return when (expanded.length) {
            6 -> {
                val r = expanded.substring(0, 2).toInt(16)
                val g = expanded.substring(2, 4).toInt(16)
                val b = expanded.substring(4, 6).toInt(16)
                (0xFF shl 24) or (r shl 16) or (g shl 8) or b
            }
            8 -> {
                val looksLikeAarrggbb = expanded.startsWith("ff") || expanded.startsWith("00")
                if (looksLikeAarrggbb) {
                    val a = expanded.substring(0, 2).toInt(16)
                    val r = expanded.substring(2, 4).toInt(16)
                    val g = expanded.substring(4, 6).toInt(16)
                    val b = expanded.substring(6, 8).toInt(16)
                    (a shl 24) or (r shl 16) or (g shl 8) or b
                } else {
                    val r = expanded.substring(0, 2).toInt(16)
                    val g = expanded.substring(2, 4).toInt(16)
                    val b = expanded.substring(4, 6).toInt(16)
                    val a = expanded.substring(6, 8).toInt(16)
                    (a shl 24) or (r shl 16) or (g shl 8) or b
                }
            }
            else -> throw IllegalArgumentException("Unsupported color format: $value")
        }
    }

    private fun argbInt(r: Double, g: Double, b: Double, a: Double): Int {
        val rr = (r * 255.0).roundToInt().coerceIn(0, 255)
        val gg = (g * 255.0).roundToInt().coerceIn(0, 255)
        val bb = (b * 255.0).roundToInt().coerceIn(0, 255)
        val aa = (a * 255.0).roundToInt().coerceIn(0, 255)
        return (aa shl 24) or (rr shl 16) or (gg shl 8) or bb
    }
}

